ldelf/ta_elf.c

```diff
@@ -262,6 +265,55 @@ static void check_hashtab(struct ta_elf *elf, void *ptr, size_t num_buckets,
 	check_range(elf, "DT_HASH", ptr, sz);
 }
 
+static void check_gnu_hashtab(struct ta_elf *elf, void *ptr)
+{
+	struct gnu_hashtab *h = ptr;
+	uint32_t *bucket = NULL;
+	size_t last_sym = 0;
+	size_t num_words = 4; /* nbuckets, symoffset, bloom_size, bloom_shift */
+	size_t bloom_words = 0;
+	size_t sz = 0;
+	size_t i = 0;
+
+	if (!IS_ALIGNED_WITH_TYPE(ptr, uint32_t))
+		err(TEE_ERROR_BAD_FORMAT, "Bad alignment of DT_GNU_HASH %p",
+		    ptr);
+
+	check_range(elf, "DT_GNU_HASH", h, sizeof(*h));
+
+	if (elf->is_32bit)
+		bloom_words = h->bloom_size;
+	else
+		bloom_words = h->bloom_size * 2;
+
+	if (ADD_OVERFLOW(num_words, h->nbuckets, &num_words) ||
+	    ADD_OVERFLOW(num_words, bloom_words, &num_words) ||
+	    MUL_OVERFLOW(num_words, sizeof(uint32_t), &sz))
+		err(TEE_ERROR_BAD_FORMAT, "DT_GNU_HASH overflow");
+
+	/*
+	 * There is no total number of exported symbols in DT_GNU_HASH,
+	 * so we have to find the highest bucket value manually.
+	 * See http://deroko.phearless.org/dt_gnu_hash.txt.
+	 */
+	bucket = (uint32_t *)h + 4 + bloom_words;
+	check_range(elf, "DT_GNU_HASH", bucket, h->nbuckets * sizeof(uint32_t));
+	last_sym = 0;
+	for (i = 0; i < h->nbuckets; i++)
+		if (bucket[i] > last_sym)
+			last_sym = bucket[i];
+
+	if (MUL_OVERFLOW(last_sym, sizeof(uint32_t), &last_sym) ||
+	    ADD_OVERFLOW(num_words, last_sym, &num_words))
```

Comment:
According to referred http://deroko.phearless.org/dt_gnu_hash.txt, `last_sym` at this stage gives the (relative) index of the last chain start, but does give last chain end. I seems to me `check_range()` below should consider the last chain size (until bit0 of a chain word is set).

Reply:
The problem is we don't know the end of the chain at this point so we can't iterate (calling `check_range()` to make sure data can be accessed **after** accessing them would not make sense of course). I am not sure how useful these range checks really are. We could either ignore this part of the check, or (better) obtain the size of the GNU hash section by other means and check against it during symbol resolution. In fact I think I can simplify the code in `save_hashtab()` by getting hash table info from the section headers (which include start and size info) rather than from the dynamic section (which contains only the start).